docs/runtime-code-basics.md

```diff
@@ -0,0 +1,229 @@
+
+| Field | Purpose |
+| ----- | ------- |
+| `context.Env` | A table of key/value pairs which are defined in the YAML [configuration](install-configuration.md) of the server. This is useful to store API keys and other secrets which may be different between servers run in production and in development. |
+| `context.ExecutionMode` | The mode associated with the execution context. It's one of these values: "after", "before", "http", or "rpc". |
+| `context.UserHandle` | The user handle associated with the execution context. It will always be `nil` on `register_http`. |
+| `context.UserId` | The user ID associated with the execution context. It will always be `nil` on `register_http`. |
+| `context.UserSessionExp` | The user session expiry in milliseconds associated with the execution context. It will always be `nil` on `register_http`. |
+
+There are four ways to register a function within the runtime each of which is used to handle specific behavior between client and server.
+
+### register_before
+
+Any function may be registered to intercept a message received from a client and operate on it (or reject it) based on custom logic. This is useful to enforce specific rules on top of the standard features in the server.
+
+```lua hl_lines="10"
+local nk = require("nakama")
+
+local function limit_friends(context, payload)
+  local user = nk.user_fetch_id({context.UserId})[1]
+  -- lets assume we've stored a user's level in their metadata.
+  local metadata = nk.json_decode(user.Metadata)
+  if metadata.level <= 10 then
+    error("Must reach level 10 before you can add friends.")
+  end
+  return payload -- important!
+end
+nk.register_before(limit_friends, "TFriendAdd")
+```
+
+The code above fetches the current user's profile and checks the metadata which is assumed to be JSON encoded with `"{level: 12}"` in it. If a user's level is too low an error is thrown to prevent the Friend Add message from being passed onwards in the server pipeline.
+
+!!! Note
+    You must remember to pass onwards the payload to the rest of the message pipeline. See `"return payload"` highlighted in the code above.
+
+### register_after
+
+Similar to [`"register_before"`](#register_before) you can attach a function to operate on a message. The registered function will be called after the message has been processed in the pipeline. The custom code will be executed asynchronously after the response message has been sent to a client.
+
+```lua
+local nk = require("nakama")
+
+local function add_reward(context, payload)
+  local value = {
+    user_ids = {payload.UserId}
+  }
+  local record = {
+    Bucket = "mygame",
+    Collection = "rewards",
+    Record = "reward",
+    UserId = context.UserId,
+    Value = value
+  }
+  nk.storage_write({ record })
+end
+
+nk.register_after(add_reward, "TFriendAdd")
+```
+
+The simple code above writes a record to a user's storage when they add a friend.
```

Comment:
I think you can add a note here that says anything data returned with be discarded and it is safe not to return anything